modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java

```diff
@@ -473,10 +473,8 @@ private JSONArray _getFiltersJSONArray(
 							WebKeys.THEME_DISPLAY);
 
 					ListTypeDefinition listTypeDefinition =
-						_listTypeDefinitionLocalService.
-							getListTypeDefinitionByExternalReferenceCode(
-								listTypeDefinitionId,
-								themeDisplay.getCompanyId());
+						_listTypeDefinitionLocalService.getListTypeDefinition(
+							listTypeDefinitionId);
```

Comment:
In difference to [the partial solution](https://github.com/liferay/liferay-portal/compare/master...dsanz:liferay-portal:LPS-196582), I opted here to use "getByID" query, instead of changing DSM persistence to store ERC. I believe that queries by ERC only make sense if we don't know the virtual instance, or want to reference any / external VI. But here, we have `themeDisplay.getCompanyId()`, currently active VI. In this case, I believe there is no difference between using ID and ERC. @dsanz am I missing some case here?

Reply:
Let's take this as a decision point to settle a pattern. There is a valid use case to justify choosing `erc` in favor of `id`

DB `id` are not portable across installations as they're assigned by the low level persistence services when creating an item. Any object entry whose fields store an `id` can't be just exported and then imported in a different installation as that `id` value is meaningless there. 

`ERC` are portable as they can be set and even modified, the value "travels" with the rest of the data entry in case data is batch-exported, and there are APIs (both REST and Java-based) to retrieve items by `ERC`. Thoug we are focusing a lot on the DSM UI, the case of using REST to batch import data sets is also worth considering, particularly when different environments are used for dev, uat and production. In this scenario, we'd like data sets to be transferrable along with any other required data/configurations across installations.

So far, `listTypeDefinitionId` is the only DB `id` we are storing as a field in an object definition. All connections between object entries (kinda FDSView → FDSEntry and so on) are internally managed by the object infra using ERC so connections between object entries will be there if we import a bunch of them on a different installation.

In light of this, I'd seriously consider using the `ERC` to store references to other stuff in the portal, as long as there is an API to retrieve items by ERC. That's the case for Picklist, which have an `ERC`. We still need to check this for vocabularies but that's a different story.

Although we're not doing it in the dynamic filters (we are fetching all picklist options from the backend and send them as `items[]`), in the event that FDS needs an `id` to either craft an URL or use it as parameter to retrieve data, we can always provide the right local value to that installation: the fragment retrieves the object by `ERC` then provides the `id` to the FDS config. As `companyId` is always implied in the rendering context, we'll use the right value.

Reply:
Got it, thanks for the explanation! :heart: 

I'll change it.

Reply:
Very good point :+1:

Reply:
> DB id are not portable across installations as they're assigned by the low level persistence services when creating an item. Any object entry whose fields store an id can't be just exported and then imported in a different installation as that id value is meaningless there.

@dsanz This part is what makes ERCs so unintuitive to me. I would expect that IDs, along with any other unique identifiers, **are portable**. The fact that IDs are typically auto-generated when you create a new entry should not be relevant. Once a unique identifier is set, it should be set. For example, my bank account has a unique identifying number. It should be the same in the live bank DB, as well as in the backup. Database entries have indexes, and they are different between installations, but that's fine. Indexes are not unique identifiers.

Is this a quirk specific to Objects backend, or am I missing something here?

Reply:
afaik, internal `id` in Liferay persistence have always been generated this way. Actually, the portability "problem" comes from the fact that moving data between installations is not a full DB backup process, as one could expect, but rather a kind of "publication" process which has 2 characteristics:
* Is partial by nature, i.e. affects a site or part of it
* It's not based on DB transfer but it's high-level, i.e. is based on full-fledged service layer where entities are sent to a different installation and "copied" there. 
 
This publication process (which we used to call staging) re-creates the things and that entails new ids, hence the non-portability thing. That's a perhaps questionable design decision, but this is how it works

Before `erc` came into play, we had the `uuid` which kinda played that role, so publication process used `uuid` to find the element to update in the target installation in case several publications take place over time. Difference perhaps is that `erc` is now made explicit and modifiable.

A quick search reveals most portal entities now have `erc` as part of the model, and so the corresponding Service APIs provide methods to fetch data by erc.